src/gpu/kmeans/kmeans_init.cu

```diff
@@ -0,0 +1,439 @@
+
+  return distance_pairs_;
+}
+
+// ArgMin operation that excludes 0. Used when dealing with distance_pairs
+// in recluster where distance between points with itself is calculated,
+// hence the name.
+// FIXME: Maybe generalize it to selection algorithm.
+template <typename T>
+struct SelfArgMinOp {
+  KmMatrix<int> argmin(KmMatrix<T>& _val) {
+    KmMatrix<int> _res(_val.rows(), 1);
+    kernel::self_row_argmin_sequential<<<
+        div_roundup(_val.rows(), 256), 256>>>(_res.k_param(),
+                                              _val.k_param());
+    return _res;
+  }
+};
+
+// MinOp that adds ESP to 0 value.
+template <typename T>
+struct SelfMinOp {
+  KmMatrix<T> min(KmMatrix<T>& _val, KmMatrixDim _dim) {
+    size_t blocks = GpuInfo::ins().blocks(32);
+    KmMatrix<T> _res(_val.rows(), 1);
+    kernel::self_row_min_sequential<<<div_roundup(_val.rows(), 256), 256>>>(
+        _res.k_param(), _val.k_param());
+    return _res;
+  }
+};
+
+// We use counting to construct the weight as described in the paper. Counting
+// is performed by histogram algorithm.
+// For re-cluster, the paper suggests using K-Means++, but that will require
+// copying data back to host. So we simply use those selected centroids with
+// highest probability.
+template <typename T>
+KmMatrix<T> GreedyRecluster<T>::recluster(KmMatrix<T>& _centroids, size_t _k) {
+
+  // Get the distance pairs for centroids
+  KmMatrix<T> centroids_dot (_centroids.rows(), 1);
+  VecBatchDotOp<T>().dot(centroids_dot, _centroids);
+  KmMatrix<T> distance_pairs (_centroids.rows(), _centroids.rows());
+  PairWiseDistanceOp<T> centroids_distance_op(
+      centroids_dot, centroids_dot, distance_pairs);
+
+  distance_pairs = centroids_distance_op(_centroids, _centroids);
+
+  // get the closest x_j for each x_i in centroids.
+  KmMatrix<int> min_indices = SelfArgMinOp<T>().argmin(distance_pairs);
+
+  // use historgram to get counting for weights
+  KmMatrix<int> weights (1, _centroids.rows());
+
+  size_t temp_storage_bytes = 0;
+  void *d_temp_storage = NULL;
+
+  // determine the temp_storage_bytes
+  safe_cuda(cub::DeviceHistogram::HistogramEven(
+      d_temp_storage, temp_storage_bytes,
+      min_indices.dev_ptr(),
+      weights.dev_ptr(),
+      min_indices.rows() + 1,
+      (T)0.0,
+      (T)min_indices.rows(),
+      (int)_centroids.rows()));
+
+  safe_cuda(cudaMalloc((void**)&d_temp_storage, temp_storage_bytes));
+  safe_cuda(cub::DeviceHistogram::HistogramEven(
+      d_temp_storage, temp_storage_bytes,
+      min_indices.dev_ptr(),    // d_samples
+      weights.dev_ptr(),        // d_histogram
+      min_indices.rows() + 1,   // num_levels
+      (T)0.0,                   // lower_level
+      (T)min_indices.rows(),    // upper_level
+      (int)_centroids.rows())); // num_samples
+  safe_cuda(cudaFree(d_temp_storage));
+
+  // Sort the indices by weights in ascending order, then use those at front
+  // as result.
+  thrust::sort_by_key(thrust::device,
```

Comment:
Be aware that this function makes a very large temporary memory allocation. Sometimes it can be better to call cub directly because of this. Another way of dealing with this is to create a custom allocator for thrust algorithms that reuses the same memory buffer.

Reply:
I have been thinking about custom memory management facility. But that might not be necessary for this function alone. Since it sorts the sampled centroids as one of the final steps. Sampled centroids should be much less then data points. If in practice this assumption doesn't hold, then kmeans|| algorithm might be a bust.